### PR TITLE
Add transpiler for multi devices

### PR DESCRIPTION
**Pull request by reyoung**
* Also make parallel_do support places

### CODE REVIEW

File: paddle/framework/init.cc [hunk 1]
```diff
@@ -77,7 +77,7 @@ bool InitDevices(const std::vector<std::string> &devices) {
 }
 
 void InitGLOG(const std::string &prog_name) {
-  google::InitGoogleLogging(prog_name.c_str());
```
Comment (reyoung) [inline paddle/framework/init.cc]: glog do not store the program name. Must use strdup to create a new buffer.

File: paddle/operators/parallel_do_op.cc [hunk 2]
```diff
@@ -234,8 +233,10 @@ class ParallelDoGradOpDescMaker : public framework::SingleGradOpDescMaker {
     auto *grad = new framework::OpDesc();
     grad->SetType("parallel_do_grad");
     for (auto &input_param : this->InputNames()) {
-      VLOG(3) << input_param;
       grad->SetInput(input_param, this->Input(input_param));
+      if (input_param == kPlaces) {
```
Comment (reyoung) [inline paddle/operators/parallel_do_op.cc]: Places do not have gradient.

File: python/paddle/v2/fluid/framework.py [hunk 3]
```diff
@@ -271,6 +271,21 @@ def _unique_var_name_():
         uid = core.unique_integer(prefix)  # unique during whole process.
         return "_".join([prefix, str(uid)])
 
+    def _ctor_param_names_(self):
+        return [
+            'type', 'name', 'shape', 'dtype', 'lod_level', 'presistable',
+            'error_clip', 'stop_gradient'
+        ]
+
+    def to_ctor_kwargs(self):
```
Comment (reyoung) [inline python/paddle/v2/fluid/framework.py]: Used by `Block::clone_var`

**Review by wangkuiyi — CHANGES_REQUESTED** [review]
Thanks for this PR. Is there a design doc for it? I am not sure do we need a transpiler for multiple devices.

**Comment by reyoung** [comment]
> Is there a design doc for it? I am not sure do we need a transpiler for multiple devices.

The transpiler is useful for multi-devices unittests. We can use the same configuration and use transpiler to transpile it to multi-devices and test the outputs are exactly the same.

If we do not implement a multi-device transpiler before, the unittests must write two methods which are basically the same.

If we are not sure whether we need a transpiler for multiple devices, I can direct implmenet the unittests without transpilers. This PR can be holded now.

File: python/paddle/v2/fluid/transpilers/multi_devices.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+def _var_lists_(var_names, block):
+    return [block.var(var_name) for var_name in var_names]
+
+
+def transpile_to_multi_devices(program,
+                               input_vars,
+                               output_vars=None,
+                               block_id=0,
+                               device_type='CPU',
+                               device_count=0):
```
Comment (tonyyang-svail) [inline python/paddle/v2/fluid/transpilers/multi_devices.py]: is `data_layer` in `parallel_do`'s `sub_block`?

File: python/paddle/v2/fluid/transpilers/multi_devices.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+
+def _copy_var_to_block_(var_names, src_block, dst_block):
+    _move_copy_var_to_block_(var_names, src_block, dst_block, remove=False)
+
+
+def _var_lists_(var_names, block):
+    return [block.var(var_name) for var_name in var_names]
+
+
+def transpile_to_multi_devices(program,
```
Comment (typhoonzero) [inline python/paddle/v2/fluid/transpilers/multi_devices.py]: We have several transpilers, are we going to all implement as a function or add some base class?

**Pull request closed without merging**